src/modules/fancyzones/lib/FancyZonesData.cpp

```diff
@@ -0,0 +1,609 @@
+    {
+        std::scoped_lock lock{ dataLock };
+        const auto& deletedCustomZoneSets = JSONHelpers::ParseDeletedCustomZoneSetsFromTmpFile(tmpFilePath);
+        for (const auto& zoneSet : deletedCustomZoneSets)
+        {
+            customZoneSetsMap.erase(zoneSet);
+        }
+    }
+
+    json::JsonObject FancyZonesData::GetPersistFancyZonesJSON()
+    {
+        return JSONHelpers::GetPersistFancyZonesJSON(zonesSettingsFileName, appZoneHistoryFileName);
+    }
+
+    void FancyZonesData::LoadFancyZonesData()
+    {
+        if (!std::filesystem::exists(zonesSettingsFileName))
+        {
+            MigrateCustomZoneSetsFromRegistry();
+
+            SaveFancyZonesData();
+        }
+        else
+        {
+            json::JsonObject fancyZonesDataJSON = GetPersistFancyZonesJSON();
+
+            appZoneHistoryMap = JSONHelpers::ParseAppZoneHistory(fancyZonesDataJSON);
+            deviceInfoMap = JSONHelpers::ParseDeviceInfos(fancyZonesDataJSON);
+            customZoneSetsMap = JSONHelpers::ParseCustomZoneSets(fancyZonesDataJSON);
+        }
+    }
+
+    void FancyZonesData::SaveFancyZonesData() const
+    {
+        std::scoped_lock lock{ dataLock };
+        json::JsonObject root{};
+        json::JsonObject appZoneHistoryRoot{};
+
+        appZoneHistoryRoot.SetNamedValue(L"app-zone-history", JSONHelpers::SerializeAppZoneHistory(appZoneHistoryMap));
+        root.SetNamedValue(L"devices", JSONHelpers::SerializeDeviceInfos(deviceInfoMap));
+        root.SetNamedValue(L"custom-zone-sets", JSONHelpers::SerializeCustomZoneSets(customZoneSetsMap));
+
+        auto before = json::from_file(zonesSettingsFileName);
+        if (!before.has_value() || before.value().Stringify() != root.Stringify())
+        {
+            Trace::FancyZones::DataChanged();
+        }
+
+        json::to_file(zonesSettingsFileName, root);
+        json::to_file(appZoneHistoryFileName, appZoneHistoryRoot);
+    }
+
+    void FancyZonesData::MigrateCustomZoneSetsFromRegistry()
+    {
+        std::scoped_lock lock{ dataLock };
+        wchar_t key[256];
+        StringCchPrintf(key, ARRAYSIZE(key), L"%s\\%s", REG_SETTINGS, L"Layouts");
+        HKEY hkey;
+        if (RegOpenKeyExW(HKEY_CURRENT_USER, key, 0, KEY_ALL_ACCESS, &hkey) == ERROR_SUCCESS)
+        {
+            BYTE data[256];
+            DWORD dataSize = ARRAYSIZE(data);
+            wchar_t value[256]{};
+            DWORD valueLength = ARRAYSIZE(value);
+            DWORD i = 0;
+            while (RegEnumValueW(hkey, i++, value, &valueLength, nullptr, nullptr, reinterpret_cast<BYTE*>(&data), &dataSize) == ERROR_SUCCESS)
+            {
+                FancyZonesDataTypes::CustomZoneSetData zoneSetData;
+                zoneSetData.name = std::wstring{ value };
+                zoneSetData.type = static_cast<FancyZonesDataTypes::CustomLayoutType>(data[2]);
+                // int version =  data[0] * 256 + data[1]; - Not used anymore
```

Comment:
Do we need this comment?

Reply:
It explains the `data[2]` above. Either I move it before it or delete it. It's fine either way

Reply:
Will remove it

Reply:
And add comment with links as discussed below